packages/compass-indexes/src/modules/create-index.tsx

```diff
@@ -0,0 +1,672 @@
+  name: O;
+  enabled: boolean;
+};
+
+export function optionToggled<O extends OptionNames>(
+  optionName: O,
+  enabled: boolean
+): OptionToggledAction<O> {
+  return { type: ActionTypes.OptionToggled, name: optionName, enabled };
+}
+
+const INITIAL_OPTIONS_STATE = Object.fromEntries(
+  (Object.keys(OPTIONS) as OptionNames[]).map((name) => {
+    return [
+      name,
+      { value: OPTIONS[name].type === 'checkbox' ? false : '', enabled: false },
+    ];
+  })
+) as Options;
+
+// other
+
+export type State = {
+  // modal state
+  inProgress: boolean;
+  isVisible: boolean;
+
+  // validation
+  error: string | null;
+
+  // form fields related
+  fields: Field[];
+
+  // index options
+  options: Options;
+};
+
+export const INITIAL_STATE: State = {
+  inProgress: false,
+  isVisible: false,
+  error: null,
+  fields: INITIAL_FIELDS_STATE,
+  options: INITIAL_OPTIONS_STATE,
+};
+
+//-------
+
+export const createIndexOpened = () => ({
+  type: ActionTypes.CreateIndexOpened,
+});
+
+export const createIndexClosed = () => ({
+  type: ActionTypes.CreateIndexClosed,
+});
+
+const errorEncountered = (error: string): ErrorEncounteredAction => ({
+  type: ActionTypes.ErrorEncountered,
+  error,
+});
+
+export const errorCleared = (): ErrorClearedAction => ({
+  type: ActionTypes.ErrorCleared,
+});
+
+const indexCreationStarted = (
+  inProgressIndex: InProgressIndex
+): IndexCreationStartedAction => ({
+  type: ActionTypes.IndexCreationStarted,
+  inProgressIndex,
+});
+
+const indexCreationSucceeded = (
+  inProgressIndexId: string
+): IndexCreationSucceededAction => ({
+  type: ActionTypes.IndexCreationSucceeded,
+  inProgressIndexId,
+});
+
+const indexCreationFailed = (
+  inProgressIndexId: string,
+  error: string
+): IndexCreationFailedAction => ({
+  type: ActionTypes.IndexCreationFailed,
+  inProgressIndexId,
+  error,
+});
+
+export type CreateIndexSpec = {
+  [key: string]: string | number;
+};
+const prepareIndex = ({
+  ns,
+  name,
+  spec,
+}: {
+  ns: string;
+  name?: string;
+  spec: CreateIndexSpec;
+}): InProgressIndex => {
+  const inProgressIndexId = new ObjectId().toHexString();
+  const inProgressIndexFields = Object.keys(spec).map((field: string) => ({
+    field,
+    value: spec[field],
+  }));
+  const inProgressIndexName =
+    name ||
+    Object.keys(spec).reduce((previousValue, currentValue) => {
+      return `${
+        previousValue === '' ? '' : `${previousValue}_`
+      }${currentValue}_${spec[currentValue]}`;
+    }, '');
+  return {
+    id: inProgressIndexId,
+    extra: {
+      status: 'inprogress',
+    },
+    key: spec,
+    fields: inProgressIndexFields,
+    name: inProgressIndexName,
+    ns,
+    size: 0,
+    relativeSize: 0,
+    usageCount: 0,
+  };
+};
+
+function isEmptyValue(value: unknown) {
+  if (value === '') {
+    return true;
+  }
+  if (value === false) {
+    return true;
+  }
+
+  return false;
+}
+
+// TODO: arguably this should move to the root store
```

Comment:
I keep going back and forth on this. Also: For search indexes we have the create/update forms' state as part of the searchIndexes slice and for regular indexes we have the create index state as a separate slice next to regularIndexes. I'm probably totally overthinking this.

Reply:
Yeah, you can kinda do it either way I think. Redux recommends modelling slices based on "[major data types or areas of functionality](https://redux.js.org/style-guide#organize-state-structure-based-on-data-types-not-components)", which if you take into account that our stores are smaller scale than the usual "whole app" can apply to the form itself if you count it as "area of functionality", so you can also argue that search index form state also deserves its own slice in this case.

Also worth mentioning, form state that can be disposed of right after the form is submitted can also live directly in UI (that's like one of the cases for not using redux that they mention a lot in docs), but from experience this works well usually only for simple-ish cases and I'm not sure if this will apply well for indexes. At least we don't really need to change this particular part right now

Reply:
I think I'll just leave it for now then. If we move it after this it should be a little bit cleaner/easier to move than before.

I did also wonder if the form state really has to be in redux, but I also remember us moving things to redux in the past so figured at least it is done already.

Reply:
Yeah, we usually do it because usually the modal toggling actions and state also pass through the store and so it makes it easier to control the clean-up instead doing it through effects in UI. Can still be done though, just kinda tricky, a very "React way" of doing it might be to just have a unique incremental "form id" in the state and use it to re-mount the form to facilitate the state reset when id changes through the `key` prop, then you still have form reset (sort of) logic in the store, but can move form input state to UI completely